presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java

```diff
@@ -285,18 +285,27 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
         @Override
         public Result apply(ValuesNode valuesNode, Captures captures, Context context)
         {
+            if (valuesNode.getRows().isEmpty()) {
+                return Result.empty();
+            }
+
             boolean anyRewritten = false;
-            ImmutableList.Builder<List<Expression>> rows = ImmutableList.builder();
-            for (List<Expression> row : valuesNode.getRows()) {
-                ImmutableList.Builder<Expression> newRow = ImmutableList.builder();
-                for (Expression expression : row) {
-                    Expression rewritten = rewriter.rewrite(expression, context);
-                    if (!expression.equals(rewritten)) {
-                        anyRewritten = true;
-                    }
-                    newRow.add(rewritten);
+            ImmutableList.Builder<Expression> rows = ImmutableList.builder();
+            for (Expression row : valuesNode.getRows().get()) {
+                Expression rewritten;
+                if (row instanceof Row) {
+                    // preserve the structure of row
+                    rewritten = new Row(((Row) row).getItems().stream()
+                            .map(item -> rewriter.rewrite(item, context))
+                            .collect(toImmutableList()));
+                }
```

Comment:
Yes... I found I needed to extract the case of Row expression because of `SimplifyExpressions`.
It uses the `LiteralEncoder` which turns a Row into something else. I had to process the Row by field and then put it back together so that it remained a `Row` at the top level.